### PR TITLE
GEO: Add support for z values to libs/geo classes

### PR DESCRIPTION
**Pull request by imotov**
Adds support for z-values to all Geometry objects in the
libs/geo library.

I marked it as WIP because of 2 outstanding issues that I would like to discuss:

1) I didn't add any support for Z serialization for rectangles. We don't support it currently on the server side, so I am not sure which format we need to adopt. Right now we support `minX, maxX, maxY, minY`. Not really sure where `minZ` and `maxZ` should go in this setup. 
2) Support for mixed dimensions in collections (multipoints, multipolygons, shells vs holes in polygons) where some elements have Z dimensions and others don't. What should we accept and what not?

### CODE REVIEW

**Comment by nknize** [comment]
> 1. I didn't add any support for Z serialization for rectangles. We don't support it currently on the server side, so I am not sure which format we need to adopt. Right now we support `minX, maxX, maxY, minY`. Not really sure where `minZ` and `maxZ` should go in this setup.

+1  This is a pain point. I have not found much in the OGC specification that thoroughly covers the `BBox` format. WKT and [GeoJSON](https://tools.ietf.org/html/rfc7946#section-5) hit on it really only from an extent perspective. The crazy `maxY` before `minY` was adopted from the  [Common Query Language specification](http://docs.opengeospatial.org/is/12-168r6/12-168r6.html) and implemented here only for compatibility with the Query Language specification. While that might be something to consider I don't think it **needs** to be the deciding factor. re: Z values for BBox / Envelope, we could consider the [Simple Feature Access specification](https://www.opengeospatial.org/standards/sfa) for `Envelope` (found on page 16), the problem here is that its more like the polygon format and quite verbose:

```
Envelope ( ): Geometry — The minimum bounding box for this Geometry, returned as a Geometry. The
polygon is defined by the corner points of the bounding box [(MINX, MINY), (MAXX, MINY), (MAXX, MAXY),
(MINX, MAXY), (MINX, MINY)]. Minimums for Z and M may be added. The simplest representation of an
Envelope is as two direct positions, one containing all the minimums, and another all the maximums. In some
cases, this coordinate will be outside the range of validity for the Spatial Reference System 
```

To me the simplest approach would be to deviate from the (non-existent) specification and just add `minZ, maxZ`. And while we're at it I think now (since this hasn't yet been released) would be the time to change from `maxY, minY` to `minY, maxY`. 

> 2\. Support for mixed dimensions in collections (multipoints, multipolygons, shells vs holes in polygons) where some elements have Z dimensions and others don't. What should we accept and what not?

I don't think we should allow mixed dimensions. I'd have to check the specification but I don't think mixed dimensions are allowed?

**Comment by imotov** [comment]
@nknize I wonder if it makes sense to support `bbox` as is in the parser for backward compatibility and introduce `bbox2d` and `bbox3d` or just simply `box` or `envelope` element with more reasonable settings.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/Circle.java [hunk 1]
```diff
@@ -26,18 +26,25 @@
...
         lon = 0;
+        alt = Double.NaN;
         radiusMeters = -1;
     }
 
     public Circle(final double lat, final double lon, final double radiusMeters) {
+        this(lat, lon, radiusMeters, Double.NaN);
+    }
+
+    public Circle(final double lat, final double lon, final double radiusMeters, final double alt) {
```
Comment (iverase) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/Circle.java]: It seems more natural for me to have alt after [lat, lon] and radius the last parameter.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java [hunk 2]
```diff
@@ -31,11 +31,15 @@ private LinearRing() {
...
+
+    public LinearRing(double[] lats, double[] lons, double[] alts) {
+        super(lats, lons, alts);
         if (lats.length < 2) {
             throw new IllegalArgumentException("linear ring cannot contain less than 2 points, found " + lats.length);
         }
-        if (lats[0] != lats[lats.length - 1] || lons[0] != lons[lons.length - 1]) {
+        if (lats[0] != lats[lats.length - 1] || lons[0] != lons[lons.length - 1] || (alts != null && alts[0] != alts[alts.length - 1])) {
             throw new IllegalArgumentException("first and last points of the linear ring must be the same (it must close itself): lats[0]="
                 + lats[0] + " lats[" + (lats.length - 1) + "]=" + lats[lats.length - 1]);
```
Comment (iverase) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java]: In the error message, could we add the alts as well?

**Review by iverase — APPROVED** [review]
Maybe we should update the java docs of the different geometries and explain what to expect from this new parameter. Apart from that LGTM

**Pull request merged**